feat(test): added test for gadgets/trace_tcpconnect

mastersans opened the pull request:
fixes: #3240

- added test for gadgets/trace_tcpconnect
- the struct for expectedEntry is created with taking into account the map in bpf file.

cc @mauriciovasquezbernal @alban

burak-ok commented on gadgets/trace_tcpconnect/test/trace_tcpconnect_test.go:
```diff
@@ -0,0 +1,127 @@
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/testing/match"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/testing/utils"
+	eventtypes "github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+	"github.com/stretchr/testify/require"
```
We normally order the imports into 3 blocks:
1. golang packages
2. third party packages
3. our own packages

This test skeleton was copied probably from another test we implemented. If yes could you point out where we missed this :sweat_smile:

mastersans replied:
@burak-ok so i did take reference from many of existing test and testing blog docs but all of those have imports in order so probably I may have added "testify/require" import in later stage of writing the test 😅, I'll fix it remember this for later contributions.

burak-ok commented on gadgets/trace_tcpconnect/test/trace_tcpconnect_test.go:
```diff
@@ -0,0 +1,127 @@
+					Proto:   6,
+				},
+				Comm: "curl",
+				Uid:  0,
+				Gid:  0,
```
Since we are not using `setuidgid` here in the command I would just set them to `utils.NormalizedInt`

burak-ok commented on gadgets/trace_tcpconnect/test/trace_tcpconnect_test.go:
```diff
@@ -0,0 +1,127 @@
+			match.MatchEntries(t, match.JSONMultiObjectMode, output, normalize, expectedEntries)
+		},
+	))
+	traceTcpconnectCmd := igrunner.New("trace tcpconnect", runnerOpts...)
```
```suggestion
	traceTcpconnectCmd := igrunner.New("trace_tcpconnect", runnerOpts...)
```

burak-ok commented on gadgets/trace_tcpconnect/test/trace_tcpconnect_test.go:
```diff
@@ -0,0 +1,127 @@
+
+	Latency     uint64           `json:"latency,omitempty"`
+	SrcEndpoint utils.L4Endpoint `json:"src"`
+	DstEndpoint utils.L4Endpoint `json:"dst"`
```
Additionally we could check for the return code of the tcpconnect gadget

mastersans replied:
@burak-ok should i check raw error or just error??

burak-ok replied:
I think the raw error is good. For 0 (no error) the message is just an empty string anyway

burak-ok reviewed (commented):
Thanks for implementing the test. It looks good

I'll start the CI after comments are addressed

mastersans commented:
@burak-ok I have made requested changes, can you take a look, Thanks.!!

burak-ok commented on gadgets/trace_tcpconnect/test/trace_tcpconnect_test.go:
```diff
@@ -0,0 +1,130 @@
+	Uid  uint32 `json:"uid"`
+	Gid  uint32 `json:"gid"`
+
+	Latency     uint64           `json:"latency,omitempty"`
```
```suggestion
	Latency     uint64           `json:"latency"`
```

burak-ok commented on gadgets/trace_tcpconnect/test/trace_tcpconnect_test.go:
```diff
@@ -0,0 +1,130 @@
+				utils.NormalizeInt(&e.MntNsID)
+				utils.NormalizeString(&e.Timestamp)
+				utils.NormalizeInt(&e.Pid)
+				utils.NormalizeInt(&e.Tid)
```
Normalizing `Uid` and `Gid` is needed

mastersans replied:
@burak-ok so should i also normalize the Latency to check its existence since it depends on the --latency flag??

burak-ok replied:
Which also means that we have to do the same for `uid` and `gid`

mastersans commented:
Hey @burak-ok I have made the desired changes, i have kept ommitempty for latency so I can add the latency test too, i have also added a little comment about error_raw for future new contributor, also from my understanding the uid and gid would be 0 since we are run test with root privileges so i have updated them accordingly, here is the local test run, 
Thanks : )

![Screenshot from 2024-08-02 15-23-41](https://github.com/user-attachments/assets/3f2e49b1-451c-4a63-8ac6-9a558e0a1552)

burak-ok commented:
The failed test about `[Test ig w/ k8s (containerd)](https://github.com/inspektor-gadget/inspektor-gadget/actions/runs/10213813538/job/28361928825?pr=3258#logs)` is a flaky test.

To pass the lint check you need to run `make lint` locally. It will either format your code according to the rules or show you what is wrong.

And finally if you could squash all your commits into a single one it would be nice

mastersans commented:
@burak-ok I have squashed the commits and fixed the linter error.

mauriciovasquezbernal commented on gadgets/trace_tcpconnect/test/trace_tcpconnect_test.go:
```diff
@@ -0,0 +1,133 @@
+				DstEndpoint: utils.L4Endpoint{
+					Addr:    "127.0.0.1",
+					Version: 4,
+					Port:    utils.NormalizedInt,
```
I think we can be sure it's always 80 (as it's the default used by curl)

mastersans replied:
@mauriciovasquezbernal I have set port for destination to 80, Thanks.

mauriciovasquezbernal replied:
This looks good. Can you squash the commits again please?

mastersans replied:
@mauriciovasquezbernal I have squashed commit for this one.

mauriciovasquezbernal reviewed (commented):
Just a minor nit, other than it, LGTM!

burak-ok reviewed (approved):
Looks good :)
Thanks for your contribution
